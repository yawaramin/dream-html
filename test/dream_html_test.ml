(* Copyright 2024 Yawar Amin

   This file is part of dream-html.

   dream-html is free software: you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation, either version 3 of the License, or (at your option) any
   later version.

   dream-html is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
   details.

   You should have received a copy of the GNU General Public License along with
   dream-html. If not, see <https://www.gnu.org/licenses/>. *)

type user =
  { name : string;
    age : int option
  }

let pp_user =
  let open Fmt in
  braces
    (record ~sep:semi
       [ field "name" (fun u -> u.name) string;
         field "age" (fun u -> u.age) (option int) ])

open Dream_html.Form

let user_form =
  let+ name = ensure "Must not be empty" (( <> ) "") required "name" string
  and+ age =
    ensure "Must be > 16"
      (Option.fold ~none:true ~some:(( < ) 16))
      optional "age" int
  in
  { name; age }

let () =
  Format.printf
    "
  OK with age: %a

  OK without age: %a

  Error without name: %a

  Error with too low age and empty name: %a
  "
    pp_user
    (Result.get_ok (validate user_form ["age", "42"; "name", "Bob"]))
    pp_user
    (Result.get_ok (validate user_form ["name", "Bob"]))
    pp_error
    (Result.get_error (validate user_form []))
    pp_error
    (Result.get_error (validate user_form ["age", "10"; "name", ""]))
