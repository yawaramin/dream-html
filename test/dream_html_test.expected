

ðŸ”Ž OK with age
{ "name" = "Bob";
  "age" = Some 42;
  "accept_tos" = true;
  "permissions" = ["r"] }

ðŸ”Ž OK without age
{ "name" = "Bob";
  "age" = None;
  "accept_tos" = true;
  "permissions" = ["r"] }

ðŸ”Ž Error without name
[name, error.required]

ðŸ”Ž Error with too low age and empty name
[age, error.range; name, Must not be empty]

ðŸ”Ž Error too many permissions
[permissions, error.length]

ðŸ”Ž Error can't have permissions if not accept TOS
[permissions, error.length]

ðŸ”Ž OK multiple nested values inside form
{ "item_count" = 2;
  "items" =
   [{ "id" = "def";
      "qty" = 10;
      "discount" = 25 };
    { "id" = "abc";
      "qty" = 1;
      "discount" = 0 }] }

ðŸ”Ž Error missing required fields of nested values
[item[0].id, error.required; item[1].qty, error.required]

ðŸ”Ž Indent CSRF tag correctly
200 OK
Content-Type: text/html; charset=utf-8
Set-Cookie: *****

<form method="post" action="/">
  <input value="token-value" name="dream.csrf" type="hidden">
  <input name="id">
  <button type="submit">Add</button>
</form>



ðŸ”Ž Redirect
303 See Other
Location: /todos/abcd


